File: src/com/secretary/UpdateInstaller.java
package com.secretary.helloworld;

import android.app.DownloadManager;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import android.widget.Toast;
import android.os.Handler;
import android.os.Looper;
import java.io.File;

public class UpdateInstaller {
    private static final String TAG = "UpdateInstaller";

    public static void downloadAndInstall(Context context, String downloadUrl, String version) {
        AppLogger logger = AppLogger.getInstance(context);
        try {
            logger.info(TAG, "Starting download for version " + version);
            logger.debug(TAG, "Download URL: " + downloadUrl);

            // Download-Manager verwenden
            DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);

            Uri uri = Uri.parse(downloadUrl);
            DownloadManager.Request request = new DownloadManager.Request(uri);
            request.setTitle("AI Secretary Update");
            request.setDescription("Downloading version " + version);
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, "AISecretary-" + version + ".apk");

            long downloadId = downloadManager.enqueue(request);
            logger.info(TAG, "Download enqueued with ID: " + downloadId);

            // Use Handler to poll download status instead of BroadcastReceiver
            // This is more reliable on newer Android versions
            Handler handler = new Handler(Looper.getMainLooper());
            Runnable checkDownload = new Runnable() {
                @Override
                public void run() {
                    AppLogger logger = AppLogger.getInstance(context);

                    DownloadManager.Query query = new DownloadManager.Query();
                    query.setFilterById(downloadId);
                    Cursor cursor = downloadManager.query(query);

                    if (cursor != null && cursor.moveToFirst()) {
                        int statusIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
                        int status = cursor.getInt(statusIndex);

                        switch (status) {
                            case DownloadManager.STATUS_SUCCESSFUL:
                                logger.info(TAG, "Download completed successfully");

                                // Get APK URI
                                String uriString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
                                Uri apkUri = Uri.parse(uriString);

                                logger.info(TAG, "Starting installation for: " + uriString);
                                Toast.makeText(context, "Download complete! Starting installation...", Toast.LENGTH_LONG).show();

                                cursor.close();

                                // Install APK
                                installApk(context, apkUri);
                                break;

                            case DownloadManager.STATUS_FAILED:
                                logger.error(TAG, "Download failed");
                                Toast.makeText(context, "Download failed. Please try again.", Toast.LENGTH_LONG).show();
                                cursor.close();
                                break;

                            case DownloadManager.STATUS_RUNNING:
                                // Get download progress
                                int bytesIndex = cursor.getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR);
                                int totalIndex = cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES);
                                int bytes = cursor.getInt(bytesIndex);
                                int total = cursor.getInt(totalIndex);

                                if (total > 0) {
                                    int progress = (bytes * 100) / total;
                                    logger.debug(TAG, "Download progress: " + progress + "%");
                                }

                                cursor.close();
                                // Check again in 500ms
                                handler.postDelayed(this, 500);
                                break;

                            case DownloadManager.STATUS_PENDING:
                            case DownloadManager.STATUS_PAUSED:
                                logger.debug(TAG, "Download status: " + status);
                                cursor.close();
                                // Check again in 500ms
                                handler.postDelayed(this, 500);
                                break;
                        }
                    } else {
                        if (cursor != null) cursor.close();
                        logger.error(TAG, "Download cursor is null or empty");
                    }
                }
            };

            logger.info(TAG, "Starting download status monitoring");
            // Start checking download status after 100ms
            handler.postDelayed(checkDownload, 100);

        } catch (Exception e) {
            logger.error(TAG, "Error downloading update", e);
        }
    }

    private static void installApk(Context context, Uri apkUri) {
        AppLogger logger = AppLogger.getInstance(context);
        try {
            logger.info(TAG, "Launching APK installer intent");
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setDataAndType(apkUri, "application/vnd.android.package-archive");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            context.startActivity(intent);
            logger.info(TAG, "APK installer intent launched successfully");
        } catch (Exception e) {
            logger.error(TAG, "Error installing APK", e);
        }
    }
}

File: app/src/main/java/com/secretary/helloworld/core/network/UpdateInstaller.kt
package com.secretary.helloworld.core.network

import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.secretary.helloworld.core.logging.AppLogger

/**
 * APK download and installation via DownloadManager
 * Phase 4.5.3 Wave 4: Converted to Kotlin
 *
 * Downloads APK from URL and launches Android installer.
 * Uses Handler-based polling to monitor download status.
 */
object UpdateInstaller {
    private const val TAG = "UpdateInstaller"

    /**
     * Download APK and install when complete
     */
    fun downloadAndInstall(context: Context, downloadUrl: String, version: String) {
        try {
            AppLogger.info(TAG, "Starting download for version $version")
            AppLogger.debug(TAG, "Download URL: $downloadUrl")

            // Use DownloadManager
            val downloadManager = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager

            val uri = Uri.parse(downloadUrl)
            val request = DownloadManager.Request(uri).apply {
                setTitle("AI Secretary Update")
                setDescription("Downloading version $version")
                setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
                setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, "AISecretary-$version.apk")
            }

            val downloadId = downloadManager.enqueue(request)
            AppLogger.info(TAG, "Download enqueued with ID: $downloadId")

            // Poll download status using Handler
            // This is more reliable than BroadcastReceiver on newer Android versions
            val handler = Handler(Looper.getMainLooper())
            val checkDownload = object : Runnable {
                override fun run() {
                    val query = DownloadManager.Query().apply {
                        setFilterById(downloadId)
                    }

                    downloadManager.query(query)?.use { cursor ->
                        if (cursor.moveToFirst()) {
                            val statusIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)
                            val status = cursor.getInt(statusIndex)

                            when (status) {
                                DownloadManager.STATUS_SUCCESSFUL -> {
                                    AppLogger.info(TAG, "Download completed successfully")

                                    // Get download URI from DownloadManager (content:// URI)
                                    val downloadUri = downloadManager.getUriForDownloadedFile(downloadId)

                                    if (downloadUri != null) {
                                        AppLogger.info(TAG, "Starting installation with URI: $downloadUri")
                                        Toast.makeText(context, "Download complete! Starting installation...", Toast.LENGTH_LONG).show()
                                        installApk(context, downloadUri)
                                    } else {
                                        AppLogger.error(TAG, "Failed to get download URI")
                                        Toast.makeText(context, "Installation failed. Please install manually from Downloads.", Toast.LENGTH_LONG).show()
                                    }
                                }

                                DownloadManager.STATUS_FAILED -> {
                                    AppLogger.error(TAG, "Download failed")
                                    Toast.makeText(context, "Download failed. Please try again.", Toast.LENGTH_LONG).show()
                                }

                                DownloadManager.STATUS_RUNNING -> {
                                    // Get download progress
                                    val bytesIndex = cursor.getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR)
                                    val totalIndex = cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES)
                                    val bytes = cursor.getInt(bytesIndex)
                                    val total = cursor.getInt(totalIndex)

                                    if (total > 0) {
                                        val progress = (bytes * 100) / total
                                        AppLogger.debug(TAG, "Download progress: $progress%")
                                    }

                                    // Check again in 500ms
                                    handler.postDelayed(this, 500)
                                }

                                DownloadManager.STATUS_PENDING,
                                DownloadManager.STATUS_PAUSED -> {
                                    AppLogger.debug(TAG, "Download status: $status")
                                    // Check again in 500ms
                                    handler.postDelayed(this, 500)
                                }
                            }
                        } else {
                            AppLogger.error(TAG, "Download cursor is null or empty")
                        }
                    } ?: run {
                        AppLogger.error(TAG, "Download cursor is null")
                    }
                }
            }

            AppLogger.info(TAG, "Starting download status monitoring")
            // Start checking download status after 100ms
            handler.postDelayed(checkDownload, 100)

        } catch (e: Exception) {
            AppLogger.error(TAG, "Error downloading update", e)
        }
    }

    /**
     * Install APK using Android Intent
     */
    private fun installApk(context: Context, apkUri: Uri) {
        try {
            AppLogger.info(TAG, "Launching APK installer with URI: $apkUri")

            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(apkUri, "application/vnd.android.package-archive")
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION

                // For Android 7+ (API 24+), must use content:// URIs
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    // URI should already be a content:// URI from DownloadManager
                    AppLogger.info(TAG, "Using content URI for Android 7+")
                }
            }

            context.startActivity(intent)
            AppLogger.info(TAG, "APK installer intent launched successfully")
            Toast.makeText(context, "Opening installer...", Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            AppLogger.error(TAG, "Error installing APK", e)
            Toast.makeText(context, "Failed to open installer. Please install manually from Downloads.", Toast.LENGTH_LONG).show()
        }
    }
}
